src/util/simplify_expr2.cpp

```diff
@@ -415,6 +415,258 @@ static type2tc common_arith_op2_type(expr2tc &e, expr2tc &f)
 
 expr2tc add2t::do_simplify() const
 {
+  // X + X --> X << 1
+  if(is_symbol2t(side_1) && is_symbol2t(side_2))
+  {
+    if(to_symbol2t(side_1) == to_symbol2t(side_2))
+      return shl2tc(type, side_1, from_integer(1, type));
+  }
+
+  // (A + 1) + ~B --> A - B
+  auto simplify_1 = [](const expr2tc e, const expr2tc f) -> expr2tc {
+    auto sidecheck = to_add2t(e).side_2;
+    auto sidecheck2 = to_add2t(e).side_1;
+    if(to_constant_int2t(sidecheck).value == 1)
+    {
+      auto B = to_bitnot2t(f).value;
+      auto new_operand = sub2tc(e->type, sidecheck2, B);
+      return new_operand;
+    }
+    return expr2tc();
+  };
+
+  if(is_add2t(side_1) && is_bitnot2t(side_2))
+  {
+    auto new_operand = simplify_1(side_1, side_2);
+    if(new_operand)
+      return new_operand;
+  }
+  if(is_bitnot2t(side_1) && is_add2t(side_2))
+  {
+    auto new_operand = simplify_1(side_2, side_1);
```

Comment:
This code repetition of the case above can be avoided by conditionally swapping `side_1` and `side_2`, e.g.
```
using std::swap;
if(is_bitnot2t(side_1) && is_add2t(side_2))
  swap(side_1, side_2);
if(is_add2t(side_1) && is_bitnot2t(side_2))
{ [...] }
```
If you did that, there would also be no need for the lambda `simplify_1`, thus removing a redundant variable name and spaghetti code.

Reply:
u r right that is a good one for simplifications where it's just two symbols, but when I try to do swap(side_1,side_2) I get an argument error because swap doesn't support expr2tc types. Does esbmc have a swap method already implemented?

Reply:
Not sure whether the irep2 expressions have their own `swap()`. If not, my example falls back to [`std::swap`](https://en.cppreference.com/w/cpp/algorithm/swap) that works with arbitrary types. You might be getting an error related to constness of the operands.